### DISCUSSION TITLE
preserve http protocol using KongIngress

### DISCUSSION THREAD

**Discussion by zzzz465** [Kong/kong#8417] [Help] [answered]
hello.

I would like to route HTTP (http/1.1, h2c, grpc plaintext) requests to backend server, while preserving HTTP version without TLS (HTTP/1.1, HTTP/2).
is there any way to make this working without using tcp passthrough?
I tried to use `KongIngress` to set `Route`, `Upstream`, but Upstream only support one protocol setting.

below is a diagram about of what I'm currently trying to achieve.

![image](https://user-images.githubusercontent.com/46273764/154015377-ce041ba3-e911-4b1f-b9fc-564bf17579ee.png)

**Comment by mayocream** [comment, marked as answer]
Hi @zzzz465, Thanks for describing your scenario clearly! 
 
Kong currently not supported to preserve the HTTP version proxying to the upstream server. You can set a separate upstream with GRPC protocol, but it still not works for h2c protocol.

If you want to dive deeper, you can write a Kong plugin to achieve it with [`preread`](https://github.com/openresty/stream-lua-nginx-module#preread_by_lua_block) phase described here.
docs: https://docs.konghq.com/gateway/2.7.x/plugin-development/custom-logic/#available-contexts

**Reply by zzzz465** [reply]
thank you for your reply!
